demos/TicTacToe/Classes/TicTacToeScene.cpp

```diff
@@ -37,51 +282,211 @@ bool TicTacToe::init() {
+  if (join_game_uuid == NULL) {
+    join_game_uuid = random_string(4);
+    ref = database->GetReference("game_data").Child(join_game_uuid);
+    firebase::Future<void> fCreateGame = ref.Child("total_players").SetValue(1);
+    fCurrentPlayerIndex =
+        ref.Child("current_player_index").SetValue(kPlayerOne);
+    WaitForCompletion(fCurrentPlayerIndex, "setCurrentPlayerIndex");
+    WaitForCompletion(fCreateGame, "createGame");
+    player_index = kPlayerOne;
+    awaiting_opponenet_move = false;
+  } else {
+    ref = database->GetReference("game_data").Child(join_game_uuid);
+    auto fIncrementTotalUsers = ref.RunTransaction([](MutableData* data) {
+      auto total_players = data->Child("total_players").value();
+      if (total_players.type() == NULL) {
+        return TransactionResult::kTransactionResultSuccess;
+      }
+      int new_total_players = total_players.int64_value() + 1;
+      if (new_total_players > kNumberOfPlayers) {
+        // Callback function needs to replace Scenes
+        return TransactionResult::kTransactionResultAbort;
+      }
+      data->Child("total_players").set_value(new_total_players);
+      // Must call this if the transaction was successful.
+      return TransactionResult::kTransactionResultSuccess;
+    });
+    WaitForCompletion(fIncrementTotalUsers, "JoinGameTransaction");
+
+    player_index = kPlayerTwo;
+    awaiting_opponenet_move = true;
+  }
 
   // Creating the board sprite , setting the position to the bottom left of the
   // frame (0,0), and finally moving the anchor point from the center of the
   // image(default) to the bottom left, Vec2(0.0,0.0).
-  Sprite* board_sprite = Sprite::create(kBoardImageFileName);
+  board_sprite = Sprite::create(kBoardImageFileName);
   if (!board_sprite) {
     log("kBoardImageFileName: %s file not found.", kBoardImageFileName);
     exit(true);
   }
   board_sprite->setPosition(0, 0);
   board_sprite->setAnchorPoint(Vec2(0.0, 0.0));
 
+  // TODO(grantpostma@): Modify these numbers to be based on the extern window
+  // size & label size dimensions.
+  cocos2d::Label* game_uuid_label =
+      Label::createWithSystemFont(join_game_uuid, "Arial", 30);
+  game_uuid_label->setPosition(Vec2(40, 20));
+  board_sprite->addChild(game_uuid_label, 1);
+  waiting_label = Label::createWithSystemFont("waiting", "Arial", 30);
+  waiting_label->setPosition(Vec2(530, 20));
+
+  board_sprite->addChild(waiting_label, 1);
+  game_over_label = Label::createWithSystemFont("", "Arial", 80);
+  game_over_label->setPosition(Vec2(300, 300));
+  board_sprite->addChild(game_over_label, 1);
+
+  // total_player_listener and CurrentPlayerIndexListener listener is set up
+  // to recognise when the desired players have connected & when turns
+  // alternate
+  LogMessage("total_player_listener");
+  total_player_listener = new ExpectValueListener(kNumberOfPlayers);
+  current_player_index_listener = new SampleValueListener();
+  last_move_listener = new SampleValueListener();
```

Comment:
I would feel better if these were stored in unique_ptrs, and allocated with std::make_unique. 

Any time you have to use new or delete, that's a code smell. As soon as you write those, you should consider whether there's a better way to manage the memory. Unless you're doing something specific, like writing your own smart pointer type, there usually is.